### PR TITLE
Async torch engine

### PR DESCRIPTION
**Pull request by grimoire**
This PR adds `thread_safe` flag to TorchEngine. ReuqestManager in the torch engine is refactored with coroutine support. CPU operation would be scheduled after `stream.synchronize`, which would benefit the throughput of restful API.

## Performance

`benchmark_restful_api.py` of llama-13b on #1197

```
RPS (request per second): 5.688 req/s
RPM (request per minute): 341.308 req/min
```

This PR

```
RPS (request per second): 6.039 req/s
RPM (request per minute): 362.370 req/min
```

throughout of `benchmark_throughput.py` won't get affected.

## Requirement:

- [x] #1197

### CODE REVIEW

File: docs/en/inference/pipeline.md [hunk 1]
```diff
@@ -224,6 +224,7 @@ This class provides the configuration parameters for Pytorch backend.
 | num_cpu_blocks        | int   | Number of CPU blocks. If the number is 0, cache would be allocated according to the current environment.                              | 0           |
 | num_gpu_blocks        | int   | Number of GPU blocks. If the number is 0, cache would be allocated according to the current environment.                              | 0           |
 | adapters              | dict  | The path configs to lora adapters.                                                                                                    | None        |
+| thread_safe           | dict  | EngineInstance is threadsafe or not.                                                                                                  | False       |
```
Comment (lvhan028) [inline docs/en/inference/pipeline.md:227]: #1203 will remove this section in `pipeline.md`
So, may rollback changes to the pipeline documents.

File: lmdeploy/pytorch/engine/engine.py [hunk 2]
```diff
@@ -207,9 +264,7 @@ def _bind_request_manager(self):
 
     def _start_loop(self):
         """start loop."""
-        loop_threads = Thread(target=self.loop, daemon=True)
-        loop_threads.start()
-        return loop_threads
+        return self.req_manager.start_loop(self.async_loop)
```
Comment (RunningLeon) [inline lmdeploy/pytorch/engine/engine.py:267]: failed with opencompass inferrence:
```
2024-03-01 14:15:32,718 - lmdeploy - INFO - build CacheEngine with config:CacheConfig(block_size=64, num_cpu_blocks=128, num_gpu_blocks=1678, window_size=-1, cache_max_entry_count=0.8)
03/01 14:15:34 - OpenCompass - INFO - Start inferencing [Llama-2-7b-chat-hf/gsm8k_0]
[2024-03-01 14:15:35,194] [opencompass.openicl.icl_inferencer.icl_gen_inferencer] [INFO] Starting inference process...
^M  0%|          | 0/12 [00:00<?, ?it/s]^M  0%|          | 0/12 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "/opt/opencompass/opencompass/tasks/openicl_infer.py", line 153, in <module>
    inferencer.run()
  File "/opt/opencompass/opencompass/tasks/openicl_infer.py", line 81, in run
    self._inference()
  File "/opt/opencompass/opencompass/tasks/openicl_infer.py", line 126, in _inference
    inferencer.inference(retriever,
  File "/opt/opencompass/opencompass/openicl/icl_inferencer/icl_gen_inferencer.py", line 146, in inference
    results = self.model.generate_from_template(
  File "/opt/opencompass/opencompass/models/base.py", line 165, in generate_from_template
    return self.generate(inputs, max_out_len=max_out_len, **kwargs)
  File "/opt/opencompass/opencompass/models/lmdeploy_pytorch.py", line 97, in generate
    _results = list(
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 619, in result_iterator
    yield fs.pop().result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 437, in result
    return self.__get_result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/usr/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/opt/opencompass/opencompass/models/lmdeploy_pytorch.py", line 149, in _generate
    generator.end(session_id)
  File "/opt/lmdeploy/lmdeploy/pytorch/engine/engine.py", line 1172, in end
    return end(self.req_sender, session_id)
  File "/opt/lmdeploy/lmdeploy/pytorch/engine/engine.py", line 135, in end
    resp = req_sender.send(RequestType.END_SESSION,
  File "/opt/lmdeploy/lmdeploy/pytorch/engine/request.py", line 345, in send
    req_id = self.send_async(req_type, data)
  File "/opt/lmdeploy/lmdeploy/pytorch/engine/request.py", line 273, in send_async
    return self.batched_send_async(req_types=[req_type], data=[data])[0]
  File "/opt/lmdeploy/lmdeploy/pytorch/engine/request.py", line 265, in batched_send_async
    return asyncio.get_event_loop().run_until_complete(coro)
  File "/usr/lib/python3.8/asyncio/events.py", line 639, in get_event_loop
    raise RuntimeError('There is no current event loop in thread %r.'
RuntimeError: There is no current event loop in thread 'ThreadPoolExecutor-0_0'.
sys:1: RuntimeWarning: coroutine 'RequestSender.async_batched_send_async' was never awaited
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

Comment (grimoire) [reply lmdeploy/pytorch/engine/engine.py:267]: Fixed, if it still not working, try enable `thread_safe` flag in `PyTorchEngineConfig`

Comment (lvhan028) [reply lmdeploy/pytorch/engine/engine.py:267]: @RunningLeon Let's prioritize our inference performance. Users are very sensitive to it.

**Comment by lvhan028** [comment]
May resolve the conflicts

File: lmdeploy/pytorch/engine/engine.py (hunk 2 above)
Comment (RunningLeon) [reply lmdeploy/pytorch/engine/engine.py:268]: ok. we may need to PR to opencompass once this PR is merged.

**Comment by RunningLeon** [comment]
BTW, `thread_safe`  set to `True` works for opencompass. And the performance of llama2 chat 7b is aligned with hf.

**Review by AllentDan — COMMENTED** [review]
用高并发的脚本跑pytorch，好像会GPU利用率不高

**Comment by grimoire** [comment]
@AllentDan 什么脚本？和 `benchmark_throughput` 和 `benchmark_restful_api` 的并发方式有什么区别吗？

**Review by RunningLeon — APPROVED** [review]
LGTM

**Comment by AllentDan** [comment]
> @AllentDan 什么脚本？和 `benchmark_throughput` 和 `benchmark_restful_api` 的并发方式有什么区别吗？

https://github.com/InternLM/lmdeploy/pull/1219#issue-2160894874
这里的。main分支好像正常

**Comment by grimoire** [comment]
@AllentDan main 好像会卡住不动

**Comment by AllentDan** [comment]
那这个问题另外PR修？

**Comment by grimoire** [comment]
@AllentDan  https://github.com/InternLM/lmdeploy/blob/f81404a67295c7e90e532797008d37fe377852bf/lmdeploy/serve/async_engine.py#L291

循环的 sleep(0) 会导致 cpu 资源一直被抢占。目前我先设置成 0.1 保证正常运行，以后可以考虑用 event 或者 asyncio.queue 来实现条件等待
还有 #1219 的脚本中设置了 repetition_penalty，如果 system prompts 里有 eos，可能会导致 eos 受到惩罚永远不生成，返回一个超长的输出。

**Review by AllentDan — APPROVED** [review]
LGTM

File: lmdeploy/serve/async_engine.py [hunk 3]
```diff
@@ -288,7 +288,7 @@ async def get_generator(self, stop: bool, session_id: int):
             return self.engine.create_instance()
         # waiting no generator is available or the same session_id is running
         while self.gens_set == set() or session_id in self.running_session_ids:
-            await asyncio.sleep(0)
+            await asyncio.sleep(0.1)
```
Comment (lvhan028) [inline lmdeploy/serve/async_engine.py:291]: 这会影响 api_server with turbomind backend 的性能么？

Comment (grimoire) [reply lmdeploy/serve/async_engine.py:291]: 基本没有

**Comment by lvhan028** [comment]
Test pipeline api using default pytorch engine config. `gemma-7b-it` failed
```
-- start to test model: /workspace/models-140/Gemma/gemma-7b-it
empty kwargs
2024-03-03 13:19:27,310 - lmdeploy - INFO - input backend=pytorch, backend_config=PytorchEngineConfig(model_name='', tp=1, session_len=None, max_batch_size=128, cache_max_entry_count=0.8, eviction_type='recompute', prefill_interval=16, block_size=64, num_cpu_blocks=0, num_gpu_blocks=0, adapters=None, max_prefill_token_num=8192, thread_safe=False, download_dir=None, revision=None)
2024-03-03 13:19:27,310 - lmdeploy - INFO - input chat_template_config=None
2024-03-03 13:19:27,310 - lmdeploy - INFO - matched chat template name: gemma
2024-03-03 13:19:28,005 - lmdeploy - INFO - updated chat_template_onfig=ChatTemplateConfig(model_name='gemma', system=None, meta_instruction=None, eosys=None, user=None, eoh=None, assistant=None, eoa=None, capability=None)
2024-03-03 13:19:29,758 - lmdeploy - INFO - Checking environment for PyTorch Engine.
2024-03-03 13:19:31,027 - lmdeploy - INFO - Checking model.

Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]
Loading checkpoint shards:  25%|██▌       | 1/4 [00:01<00:03,  1.01s/it]
Loading checkpoint shards:  50%|█████     | 2/4 [00:02<00:02,  1.13s/it]
Loading checkpoint shards:  75%|███████▌  | 3/4 [00:03<00:01,  1.18s/it]
Loading checkpoint shards: 100%|██████████| 4/4 [00:03<00:00,  1.10it/s]
Loading checkpoint shards: 100%|██████████| 4/4 [00:03<00:00,  1.01it/s]2024-03-03 13:19:41,946 - lmdeploy - INFO - build CacheEngine with config:CacheConfig(block_size=64, num_cpu_blocks=146, num_gpu_blocks=1803, window_size=-1, cache_max_entry_count=0.8)
2024-03-03 13:19:47,080 - lmdeploy - INFO - updated backend_config=PytorchEngineConfig(model_name='', tp=1, session_len=2048, max_batch_size=128, cache_max_entry_count=0.8, eviction_type='recompute', prefill_interval=16, block_size=64, num_cpu_blocks=0, num_gpu_blocks=0, adapters=None, max_prefill_token_num=8192, thread_safe=False, download_dir=None, revision=None)
../aten/src/ATen/native/cuda/Indexing.cu:1239: indexSelectSmallIndex: block: [2,0,0], thread: [0,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
../aten/src/ATen/native/cuda/Indexing.cu:1239: indexSelectSmallIndex: block: [2,0,0], thread: [1,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
../aten/src/ATen/native/cuda/Indexing.cu:1239: indexSelectSmallIndex: block: [2,0,0], thread: [2,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
```

**Pull request merged**